File: cpp/62. unique-paths.cpp
//
// Created by Rutvik Patel on 1/31/24.
//

#include "include.h"

// solution
class Solution {
public:
    int uniquePaths(int m, int n) {
        if (m < n) std::swap(m, n);
        long long unique_paths = 1;

        for (int idx = 0; idx < n - 1; ++idx) {
            unique_paths *= (m + idx);
            unique_paths /= (idx + 1);
        }

        return unique_paths;
    }
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::pair<int, int>, int>> test_cases = {
        { { 3, 7 }, 28 },
        { { 3, 2 }, 3 },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.uniquePaths(args.first, args.second) == ans, "Failed!");

    log("All testcases passed!");
}

File: python/62. unique-paths.py
from typing import List


# lc sol start
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        down, right = m - 1, n - 1
        total = down + right

        dp = [1] * (total + 1)

        for idx in range(1, total + 1):
            dp[idx] = idx * dp[idx - 1]

        return dp[-1] // (dp[down] * dp[right])


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [(3, 7), (3, 2)],
        'outputs': [28, 3],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.uniquePaths(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()
